### PR TITLE
Gets rid of registerDecorationType in inlay hints controller

### PR DESCRIPTION
**Pull request by hediet**
See #132537.

The utility introduced by this PR can be made a general utility later.

### CODE REVIEW

File: src/vs/editor/contrib/inlayHints/inlayHintsController.ts [hunk 1]
```diff
@@ -209,13 +208,14 @@ export class InlayHintsController implements IEditorContribution {
 		return result;
 	}
 
+	private readonly ruleFactory = new DynamicCssRules(this._editor);
```
Comment (jrieken) [inline src/vs/editor/contrib/inlayHints/inlayHintsController.ts:211]: nit - `_` prefix for privates, also move atop ctor

File: src/vs/editor/contrib/inlayHints/utils.ts [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+		return this.referenceCount > 0;
+	}
+}
+
+function camelToDashes(str: string): string {
+	return str.replace(/(^[A-Z])/, ([first]) => first.toLowerCase())
+		.replace(/([A-Z])/g, ([letter]) => `-${letter.toLowerCase()}`);
+}
+
+function themeColorToCssVar(themeColor: ThemeColor): string {
```
Comment (jrieken) [inline src/vs/editor/contrib/inlayHints/utils.ts:151]: There is [`asCssVariableName`](https://github.com/microsoft/vscode/blob/d5db92d1f579fcc7fc1963a6baf16f39c0360833/src/vs/platform/theme/common/colorRegistry.ts#L34)

File: src/vs/editor/contrib/inlayHints/utils.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+import * as dom from 'vs/base/browser/dom';
+import { ThemeColor } from 'vs/platform/theme/common/themeService';
+
+/**
+ * A helper to create dynamic css rules, bound to a class name.
+ * Rules are reused.
+ * Reference counting and delayed garbage collection ensure that no rules leak.
+*/
+export class DynamicCssRules {
+	private counter = 0;
```
Comment (jrieken) [inline src/vs/editor/contrib/inlayHints/utils.ts:18]: nit - more `_` for privates

**Review by jrieken — APPROVED** [review]
Left some, for future please don't do "unrelated" change renaming variables because that makes review unnecessary complex.

Tho, not sure what the overall motivation is? Is before/afterInjectedText going aways?

**Comment by hediet** [comment]
> Tho, not sure what the overall motivation is? Is before/afterInjectedText going aways?

Yes. Ideally, everything related to "decoration types" is going away, as its implementation is very complicated.

**Pull request merged**